File: algorithm/data_structure/array/array.c
#include "array.h"

int init(Array *array)
{
	array->length = 0;
	return 0;
}

int at(Array *array, int k)
{
	if(k > array->length || k <= 0) return -1;
	return array->elements[k - 1];
}

int insert(Array *array, int k, int val)
{
	if(array->length >= ARRAY_MAX_SIZE || k > array->length + 1) return -1;

	for(int i = array->length; i >= k; i--) {
		array->elements[i] = array->elements[i - 1];
	}
	array->length += 1;
	array->elements[k - 1] = val;

	return 0;
}

int delete(Array *array, int k)
{
	array->length -= 1;
	for(int i = k - 1; i < array->length; i++) {
		array->elements[i] = array->elements[i+1];
	}

	return 0;
}

File: DSA/Array.java
public class Array {
    private final int MAX_SIZE = 1024;
    private int size;
    private int[] data;

    public Array() {
        this.size = 0;
        this.data = new int[MAX_SIZE];
    }

    public int size() {
        return this.size;
    }

    public int get(int index) {
        if (index >= 0 && index < this.size) {
            return this.data[index];
        } else {
            throw new IndexOutOfBoundsException("Array");
        }
    }

    public void insert(int index, int value) {
        if (index < 0 || index > this.size) {
            throw new IndexOutOfBoundsException("Array");
        }

        for (int i = this.size; i > index; i--) {
            this.data[i] = this.data[i-1];
        }
        this.data[index] = value;

        this.size += 1;
    }

    public int delete(int index) {
        if (index < 0 || index >= this.size) {
            throw new IndexOutOfBoundsException("Array");
        }

        int deleted = this.data[index];
        this.size -= 1;

        for (int i = index; i < this.size; i++) {
            this.data[i] = this.data[i+1];
        }

        return deleted;
    }

    public void display() {
        System.out.print("[");
        for (int i = 0; i < this.size - 1; i++) {
            System.out.printf("%d, ", this.data[i]);
        }
        System.out.printf("%d]\n", this.data[this.size - 1]);
    }

    public void bubble_sort() {
        int tmp;
        for (int i = 0; i < this.size - 1; i++) {
            for (int j = 0; j < this.size - i - 1; j++) {
                if (this.data[j] > this.data[j+1]) {
                    tmp = this.data[j];
                    this.data[j] = this.data[j+1];
                    this.data[j+1] = tmp;
                }
            }
        }
    }
}
